baseplate_celery/__init__.py

```diff
@@ -0,0 +1,130 @@
+import logging
+
+from baseplate import Baseplate
+from baseplate.clients.sqlalchemy import SQLAlchemySession
+from baseplate.lib import config
+from baseplate.lib.secrets import secrets_store_from_config
+from celery import Celery
+from celery import Task
+from celery.worker.request import Request
+from psycogreen.gevent import patch_psycopg
+
+logger = logging.getLogger(__name__)
+
+
+def parse_celery_configs(app_config: config.RawConfig, prefix: str = "celery."):
+    """Make a BaseplateCelery application from a configuration dictionary.
+
+    The keys useful to :py:func:`celery_from_config` should be prefixed, e.g.
+    ``celery.service_name``, ``celery.broker_url``, etc. The ``prefix`` argument
+    specifies the prefix used to filter keys.
+
+    Supported keys:
+
+    * ``service_name``          (required): the name of service.
+    * ``broker_url``            (required): a Redis URL like ``redis://localhost/0``.
+    * ``db.url``                (optional): Postgres configs for backend database.
+    * ``db.credentials_secret`` (optional): the key used to retrieve the database credentials from ``secrets``
+
+    """
+    assert prefix.endswith(".")
+    parser = config.SpecParser(
+        {
+            "service_name": config.String,
+            "broker_url": config.String,
+            "db": config.DictOf(config.String),
+        }
+    )
+    options = parser.parse(prefix[:-1], app_config)
+    return options
+
+
+def short_task_name(task_name: str) -> str:
+    return task_name.split(".")[-1]
+
+
+def emit_celery_metric(context, task_name, **tags):
+    tags = {"task_name": short_task_name(task_name), **tags}
+    context.metrics.counter("celery_task", tags=tags).increment()
+    context.metrics.flush()
+
+
+class BaseplateRequest(Request):
+    def on_timeout(self, soft, timeout):
+        super().on_timeout(soft, timeout)
+        emit_celery_metric(
+            self.task._context, self.task.name, success=False, timeout=True
+        )
+        self._context.sentry.capture_message(
+            "Task {self.task.name} timed out after {timeout} seconds"
+        )
+
+
+class BaseplateTask(Task):
+    Request = BaseplateRequest
+    acks_late = True
+    track_started = True
+
+    _context = None
```

Comment:
not being super familiar with celery, how do Task objects get instantiated? do we ever have to worry about concurrency here? if so might a task accidentally use the context from someone else?